engines/wintermute/base/gfx/opengl/base_render_opengl3d.cpp

```diff
@@ -0,0 +1,384 @@
+/* ResidualVM - A 3D game interpreter
+ *
+ * ResidualVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	 See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "base_render_opengl3d.h"
+
+#include "../../../../../graphics/opengl/system_headers.h"
+#include "base_surface_opengl3d.h"
+#include "math/glmath.h"
+
+namespace Wintermute {
+BaseRenderer *makeOpenGL3DRenderer(BaseGame *inGame) {
+	return new BaseRenderOpenGL3D(inGame);
+}
+
+} // namespace Wintermute
+
+Wintermute::BaseRenderOpenGL3D::BaseRenderOpenGL3D(Wintermute::BaseGame *inGame)
+    : _spriteBatchMode(false) {
+}
+
+Wintermute::BaseRenderOpenGL3D::~BaseRenderOpenGL3D() {
+}
+
+Wintermute::BaseImage *Wintermute::BaseRenderOpenGL3D::takeScreenshot() {
+	return nullptr;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::saveScreenShot(const Common::String &filename, int sizeX, int sizeY) {
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setViewport(int left, int top, int right, int bottom) {
+	_viewportRect.setRect(left, top, right, bottom);
+	glViewport(left, top, right - left, bottom - top);
+	return true;
+}
+
+bool Wintermute::BaseRenderOpenGL3D::setViewport(Wintermute::Rect32 *rect) {
+	return setViewport(rect->left, rect->top, rect->right, rect->bottom);
+}
+
+Wintermute::Rect32 Wintermute::BaseRenderOpenGL3D::getViewPort() {
+	return Rect32(0, 0, 0, 0);
+}
+
+void Wintermute::BaseRenderOpenGL3D::setWindowed(bool windowed) {
+}
+
+Graphics::PixelFormat Wintermute::BaseRenderOpenGL3D::getPixelFormat() const {
+	return OpenGL::Texture::getRGBAPixelFormat();
+}
+
+void Wintermute::BaseRenderOpenGL3D::fade(uint16 alpha) {
+	fadeToColor(0, 0, 0, (byte)(255 - alpha));
+}
+
+void Wintermute::BaseRenderOpenGL3D::fadeToColor(byte r, byte g, byte b, byte a) {
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+
+	setProjection2D();
+
+	int vertex_size = 16;
+	byte vertices[4 * vertex_size];
+
+	*reinterpret_cast<float *>(vertices + 4) = _viewportRect.left;
```

Comment:
Consider casting the pointer once, or possibly defining a packed struct to make this cleaner.